src/wp-admin/includes/file.php

```diff
@@ -1946,6 +1946,83 @@ function copy_dir( $from, $to, $skip_list = array() ) {
 	return true;
 }
 
+/**
+ * Moves a directory from one location to another.
+ *
+ * Recursively invalidates OPcache on success.
+ *
+ * If the renaming failed, falls back to copy_dir().
+ *
+ * Assumes that WP_Filesystem() has already been called and setup.
+ *
+ * @since 6.2.0
+ *
+ * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+ *
+ * @param string $from      Source directory.
+ * @param string $to        Destination directory.
+ * @param bool   $overwrite Optional. Whether to overwrite the destination directory if it exists.
+ *                          Default false.
+
+ * @return true|WP_Error True on success, WP_Error on failure.
+ */
+function move_dir( $from, $to, $overwrite = false ) {
+	global $wp_filesystem;
+
+	if ( trailingslashit( strtolower( $from ) ) === trailingslashit( strtolower( $to ) ) ) {
+		return new WP_Error(
+			'source_destination_same_move_dir',
+			__( 'The source and destination are the same.' )
+		);
+	}
+
+	if ( ! $overwrite && $wp_filesystem->exists( $to ) ) {
+		return new WP_Error(
+			'destination_already_exists_move_dir',
+			sprintf(
+				/* translators: %s: The destination directory path. */
+				__( 'The destination folder, %s, already exists.' ),
+				"<code>$to</code>"
+			)
+		);
+	}
+
+	if ( $wp_filesystem->move( $from, $to, $overwrite ) ) {
+		/*
+		 * When using an environment with shared folders,
+		 * there is a delay in updating the filesystem's cache.
+		 *
+		 * This is a known issue in environments with a VirtualBox provider.
+		 *
+		 * A 200ms delay gives time for the filesystem to update its cache,
+		 * prevents "Operation not permitted", and "No such file or directory" warnings.
+		 *
+		 * This delay is used in other projects, including Composer.
+		 * @link https://github.com/composer/composer/blob/2.5.1/src/Composer/Util/Platform.php#L228-L233
+		 */
+		usleep( 200000 );
+		wp_opcache_invalidate_directory( $to );
+
+		return true;
+	}
+
+	// Fall back to a recursive copy.
+	if ( ! $wp_filesystem->is_dir( $to ) ) {
+		if ( ! $wp_filesystem->mkdir( $to, FS_CHMOD_DIR ) ) {
+			return new WP_Error( 'mkdir_failed_move_dir', __( 'Could not create directory.' ), $to );
+		}
+	}
+
+	$result = copy_dir( $from, $to, array( basename( $to ) ) );
+
+	// Clear the source directory.
+	if ( ! is_wp_error( $result ) ) {
```

Comment:
```suggestion
	if ( true === $result ) {
```

`copy_dir()` would return `true` on success or WP_Error. Checking for `true` is a tiny optimization, perhaps :)

Reply:
That code is not new to the PR.

Reply:
`is_wp_error()` has the `is_wp_error_instance` action hook [ref](https://developer.wordpress.org/reference/functions/is_wp_error/).

Is this useful to have here, or is the optimization better?

Reply:
`! is_wp_error()` might be a little more explicit. While `true === $result` is clearly a successful result, `! is_wp_error()` is explicitly stating "Delete the source if an error hasn't occurred."

Reply:
Either way works here. The code in `WP_Upgrader::install_package()`, while similar, does not seem exactly the same, as it checks for `is_wp_error()` (non-negative). No strong preference from me, but checking for `true === $result` explicitly should be fine 👍